src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -331,9 +366,63 @@ public object GetOrRegisterObjectForComInstance(IntPtr externalComObject, Create
             if (flags.HasFlag(CreateObjectFlags.Aggregation))
                 throw new NotImplementedException();
 
-            object? wrapperMaybeLocal = wrapperMaybe;
-            retValue = null;
-            throw new NotImplementedException();
+            if (flags.HasFlag(CreateObjectFlags.Unwrap))
+            {
+                var comInterfaceDispatch = impl.TryGetComInterfaceDispatch(externalComObject);
+                if (comInterfaceDispatch != null)
+                {
+                    retValue = ComInterfaceDispatch.GetInstance<object>(comInterfaceDispatch);
+                    return true;
+                }
+            }
+
+            if (!flags.HasFlag(CreateObjectFlags.UniqueInstance))
+            {
+                using (LockHolder.Hold(impl._lock))
+                {
+                    if (impl._rcwCache.TryGetValue(externalComObject, out var handle))
+                    {
+                        if (handle.IsAllocated)
+                        {
+                            retValue = handle.Target;
+                            return false;
+                        }
+                    }
+                }
+            }
+
+            retValue = impl.CreateObject(externalComObject, flags);
+            if (retValue == null)
+            {
+                // If ComWrappers instance cannot create wrapper, we can do nothing here.
+                return false;
+            }
+
+            if (flags.HasFlag(CreateObjectFlags.UniqueInstance))
+            {
+                // No need to cache NativeObjectWrapper for unique instances. They are not cached.
+                return true;
+            }
+
+            using (LockHolder.Hold(impl._lock))
+            {
+                if (impl._rcwCache.TryGetValue(externalComObject, out var existingHandle))
+                {
+                    if (existingHandle.IsAllocated)
+                    {
+                        retValue = existingHandle.Target;
+                    }
+                }
+                else
+                {
+                    NativeObjectWrapper wrapper = new NativeObjectWrapper(externalComObject, impl);
+                    impl._rcwTable.Add(retValue, wrapper);
+                    impl._rcwCache.Add(externalComObject, GCHandle.Alloc(retValue, GCHandleType.Weak));
```

Comment:
With GCHandle, you will also need to cleanup all GCHandles if the ComWrappers instance is garbage collected. 

It may make sense for the GCHandle to be owned by `NativeObjectWrapper` and delete it in `NativeObjectWrapper` finalizer.

Reply:
Then it require me to `Dictionary<IntPtr, NativeObjectWrapper> _rcwCache` instead of `Dictionary<IntPtr, GCHandle> _rcwCache`. Would be finalizer for `NativeObjectWrapper` even called in that way? I use GCHandle now to break cycle
`ComProxy` -> `NativeObjectWrapper` -> `ComWrappers` -> `_rcwCache<IntPtr, GCHandle<ComProxy>>`.
If link between `ComProxy` -> `NativeObjectWrapper` is weak, and `NativeObjectWrapper` will be finalized even other parts of code have reference on it, I can do what you propose.

Reply:
I think the dictionary can still be just `Dictionary<IntPtr, GCHandle>`, even when you do this.